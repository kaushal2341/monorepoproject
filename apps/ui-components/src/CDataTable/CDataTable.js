import React, { PureComponent } from 'react'
import { AgGridReact} from 'ag-grid-react'
import 'ag-grid/dist/styles/ag-grid.css';
import 'ag-grid/dist/styles/ag-theme-balham.css';
import PropTypes from 'prop-types';
class CDataTable extends PureComponent {
  constructor (props) {
    super(props)
    this.state = {
      gridApi: '',
      rowChanged: []
    }
  }

  onGridReady = params => {
    console.log(params)  
    this.setState({
      gridApi: params
    })
  }

  checkDataAndFilterTheData = event => {
    let datum = [...this.state.rowChanged]
    for (let i = 0; i < this.props.rowData.length; i++) {
      if (datum.length > i) {
        if (datum[i].id === event.data.id) {
          datum[i] = { ...event.data }
          break
        }
      } else {
        datum[i] = { ...event.data }
        break
      }
    }
    return datum;
  }

  onCellValueChanged = e => {
    let data = this.checkDataAndFilterTheData(e)
    this.setState({
        rowChanged:data
    })
  }
   
  onSelectionChanged = (row) => {
    this.props.onSelectionChanged(row.api.getSelectedRows()[0])
  }

  render () {
    const {
      id,
      tableId,
      width,
      height,
      classes,
      enableSorting,
      columnDefs,
      rowData,
      defaultColDef,
      rowSelection,
      frameworkComponents,
      floatingFilter,
      editType
     } = this.props
    return (
      <>
        <div
          style={{ width: width, height: height }}
          className={classes}
          id={id}
        >
          <AgGridReact
            id={tableId}
            enableSorting={enableSorting}
            columnDefs={columnDefs}
            rowData={rowData}
            defaultColDef={defaultColDef}
            onGridReady={this.onGridReady}
            rowSelection={rowSelection}
            frameworkComponents={frameworkComponents}
            floatingFilter={floatingFilter}
            editType={editType}
            onSelectionChanged={(e)=>this.onSelectionChanged(e)}
            onCellValueChanged={(e)=>this.onCellValueChanged(e)}
          />
        </div>
      </>
    )
  }
}

export default CDataTable;

