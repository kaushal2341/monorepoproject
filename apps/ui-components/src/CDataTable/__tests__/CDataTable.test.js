import CDataTable from '../CDataTable'
import { AgGridReact } from 'ag-grid-react'
// import { EventService } from 'ag-grid-community';
expect.addSnapshotSerializer(enzymeSerializer)
describe('CDataTable Component Test', () => {
  let sShallow, onSelectionChangedMock, rowData, columnDefs, smount
  columnDefs = [
    { headerName: 'Make', field: 'make' },
    { headerName: 'Model', field: 'model' },
    { headerName: 'Price', field: 'price' }
  ]

  // specify the data
  rowData = [
    { make: 'Toyota', model: 'Celica', price: 35000 },
    { make: 'Ford', model: 'Mondeo', price: 32000 },
    { make: 'Porsche', model: 'Boxter', price: 72000 }
  ]

  beforeEach(() => {
    onSelectionChangedMock = jest.fn()
    sShallow = shallow(
      <CDataTable
        id='agGridTableWrapper'
        tableId='agGridTable'
        onSelectionChanged={onSelectionChangedMock}
        rowData={rowData}
        columnDefs={columnDefs}
        rowSelection='single'
      />
    )
    // console.log(CDataTable.dive())
  })

  it('should be defined', () => {
    expect(sShallow).toBeDefined()
  })
  it('should have only one table wrapper', () => {
    expect(sShallow.find('#agGridTableWrapper')).toHaveLength(1)
  })
  it('should have only one Ag grid react table component', () => {
    expect(sShallow.find('#agGridTable')).toHaveLength(1)
  })
  it('should have wrapper with given props available', () => {
    const prps = ['height', 'width', 'id']
    const prpsReq = Object.keys(sShallow.find('#agGridTableWrapper').props())
    prpsReq.map((prp, i) => {
      expect(prp).toContain(prpsReq[i])
    })
  })
  it('should have table with given props available', () => {
    const prps = [
      'id',
      'enableSorting',
      'columnDefs',
      'rowData',
      'defaultColDef',
      'onGridReady',
      'rowSelection',
      'frameworkComponents',
      'floatingFilter',
      'editType',
      'onSelectionChanged',
      'onCellValueChanged'
    ]
    const keys = Object.keys(sShallow.find('#agGridTable').props())
    prps.map((prp, i) => {
      expect(prps).toContain(keys[i])
    })
  })
  describe('Simulation Change', () => {
    let changeCallback;
    const simulation = (myEventFunction,event,value) => {
      changeCallback = jest.spyOn(sShallow.instance(), myEventFunction)
      sShallow.instance().forceUpdate()
      sShallow.find(AgGridReact).simulate(event, value)        
    }
    it('should simulate the function onSelectionChanged when table value is changed', () => {
      simulation('onSelectionChanged','selectionChanged',{
        api: { getSelectedRows: jest.fn(a => []) }
      });
      expect(onSelectionChangedMock).toHaveBeenCalled();
      expect(changeCallback).toHaveBeenCalled()
    })

    it('should simulate the function onCellValueChanged and change the state rowChanged when table value is changed', () => {
      const value = {
        data: {id:1,make: 'Porsche', model: 'Boxer', price: 72000}
      }
      simulation('onCellValueChanged','cellValueChanged',value)
      
      expect(sShallow.state().rowChanged[0]).toEqual(value.data)
      expect(changeCallback).toHaveBeenCalled()
    });

    it('should change only the value of rowChanged when same id of object is given',() => {
      let value = {
        data: {id:1,make: 'Porsche', model: 'Boxer', price: 72000}
      }
      simulation('onCellValueChanged','cellValueChanged',value)
      value = {
        data: {id:1,make: 'Porsch', model: 'Boxer', price: 72000}
      }
      simulation('onCellValueChanged','cellValueChanged',value)
      expect(sShallow.state().rowChanged[0]).toEqual(value.data);
    })
  })
  describe('Snapshot Testing', () => {
    it('should match snapshot', () => {
      expect(sShallow).toMatchSnapshot()
    })
  })
})
