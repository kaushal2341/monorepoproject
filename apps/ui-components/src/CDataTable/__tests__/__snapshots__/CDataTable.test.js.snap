// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CDataTable Component Test Snapshot Testing should match snapshot 1`] = `
<Fragment>
  <div
    id="agGridTableWrapper"
    style={
      Object {
        "height": undefined,
        "width": undefined,
      }
    }
  >
    <AgGridReact
      columnDefs={
        Array [
          Object {
            "field": "make",
            "headerName": "Make",
          },
          Object {
            "field": "model",
            "headerName": "Model",
          },
          Object {
            "field": "price",
            "headerName": "Price",
          },
        ]
      }
      id="agGridTable"
      onCellValueChanged={[Function]}
      onGridReady={[Function]}
      onSelectionChanged={[Function]}
      rowData={
        Array [
          Object {
            "make": "Toyota",
            "model": "Celica",
            "price": 35000,
          },
          Object {
            "make": "Ford",
            "model": "Mondeo",
            "price": 32000,
          },
          Object {
            "make": "Porsche",
            "model": "Boxter",
            "price": 72000,
          },
        ]
      }
      rowSelection="single"
    />
  </div>
</Fragment>
`;
